### DISCUSSION TITLE
Is there a way to prevent the background color resizing from being delayed when resizing the window?

### DISCUSSION THREAD

**Discussion by nathanAjacobs** [unoplatform/uno#8957] [Q&A]
When I resize my window the background color is delayed to fill the whole window. Is there a way to fix this?

https://user-images.githubusercontent.com/37193095/171512193-157a4576-8259-49ff-b1a3-a5e72b2ce5be.mp4

**Comment by nathanAjacobs** [comment]
This actually seems to be a problem with WinUI 3 itself. Hoping a fix is released soon, this is a dealbreaker for me.

**Reply by jeromelaban** [reply]
In general, there's a delay when re-rendering the window content after a resize, and in general, this is handled by ensuring that the window background matches a color that won't be significantly different from the app's content. 

Uno's targets also have the similar behavior, because the rendering is performed on a thread different from the windowing system's one. You'll see similar behavior on mobiles, particularly when changing screen orientations.

Also, if you're running the Skia+WPF head, the rendering is still done using software rendering, which makes it a bit slower than it could be. You can try with Skia+GTK using Uno 4.3.8, which uses OpenGL acceleration.

**Comment by nathanAjacobs** [comment]
Is it possible to change the Window background itself as you described? As far as I can tell, WinUI 3 XAML doesn't have a Background property on the Window class.

I also tried Skia+GTK, but the problem was still there.

**Reply by jeromelaban** [reply]
It's a question to ask on the [WinUI repository](https://github.com/microsoft/microsoft-ui-xaml).

For Skia+GTK, it's a similar problem to WinUI, though the API is likely to be different and we do not have one yet (Let's track it here https://github.com/unoplatform/uno/issues/8961). For Skia+WPF, you can change the XAML in the WPF host to another color.

**Comment by nathanAjacobs** [comment]
@jeromelaban Maybe I'm misunderstanding, but setting the Window or ContentControl's background via XAML in WPF host has no effect. The only thing that has an effect is setting the background color in code behind like so:

```cs
public MainWindow()
{
    InitializeComponent();

    root.Content = new global::Uno.UI.Skia.Platform.WpfHost(Dispatcher, () => new TestUnoApp.App());
    var contentControl = root.Content as Uno.UI.Skia.Platform.WpfHost;
    contentControl.Background = new SolidColorBrush(new Color { R = 0, G = 0, B = 0, A = 128 });
}
```

But even that behaves weird because it actually draws on top of the regular shared apps content. For example the shared apps main Page has it's background set to orange but as you can see it draws on top, if I change the alpha value to 255 it completely blocks the shared app and just draws black.

https://user-images.githubusercontent.com/37193095/171903412-e9879fe1-a7d4-40d4-9add-b8b48ec3d63a.mp4

**Reply by jeromelaban** [reply]
You'll likely need to change another color. In this case, you're changing the inner element (`contentControl`), but above there's `root`, and there's also the `Window`, for which there's probably another way to change the color. This is general WPF knowledge, you may find that answer in the Microsoft WPF documentation.

**Reply by nathanAjacobs** [reply]
Like I said I tried changing the root, window, and contentControl background first but it had no effect

**Reply by jeromelaban** [reply]
Got it. You'll still need to search for ways to do this with the WPF documentation, Uno does not provide anything specific to solve this issue.

**Reply by nathanAjacobs** [reply]
My point is, it's the Skia Canvas element causing this issue in WPF host.

```cs
public MainWindow()
{
    InitializeComponent();

    Background = new SolidColorBrush(Colors.Red);
    //root.Content = new global::Uno.UI.Skia.Platform.WpfHost(Dispatcher, () => new TestUnoApp.App());
}
```

If I comment out the above line, the background of the window is red, when it is uncommented (i.e. drawing the Skia portion) the background is white in the delayed resizing portion. So something in the Skia Canvas implementation is drawing over the window background completely with a white background. I'm not entirely sure, but I think it's stemming from this portion of the WpfHost class, when it clears the canvas with a white color:

```cs
// draw on the bitmap
bitmap.Lock();
using (var surface = SKSurface.Create(info, bitmap.BackBuffer, bitmap.BackBufferStride))
{
    surface.Canvas.Clear(SKColors.White);
    surface.Canvas.SetMatrix(SKMatrix.CreateScale((float)dpiScaleX, (float)dpiScaleY));
    WinUI.Window.Current.Compositor.Render(surface);
}
```

So there is actually no way for me to change that white color via pure WPF, because it is in the WpfHost's Skia rendering implementation.

**Reply by jeromelaban** [reply]
I see, but in this case the white color you're mentioning is at the base of the Skia canvas, and this one can be changed with the background of the first element of the Uno tree. This particular part is not asynchronous, and if the first element draws its background, the White color will not be visible. If you're seeing another color, it's coming from WPF.

**Reply by nathanAjacobs** [reply]
I tried changing the canvas (first element in the uno tree) background color last night as well. Unfortunately, it just completely renders over the content from the shared app. So If I set the canvas background color to black, I just have a completely black window.

**Reply by jeromelaban** [reply]
Would you be able to post a small repro app which shows the issues your getting in the video?

**Reply by nathanAjacobs** [reply]
Yes, I can do this later tonight when I get home from work. I will update you later when it is up.

**Reply by nathanAjacobs** [reply]
Here is the demo WPF project which demonstrates this issue:

https://github.com/nathanAjacobs/UnoAppBackgroundBugWPF

In the MainWindow.xaml.cs file of the WPF host project, there are several commented out lines which sets various `Background` colors on the corresponding elements. If you uncomment the lines, you should see the behavior of it rendering the background over the shared app's content.

**Reply by jeromelaban** [reply]
Thanks for the repro. For Skia Gtk/WPF, this PR should help: https://github.com/unoplatform/uno/pull/9039.
